src/poetry/console/commands/self/self_command.py

```diff
@@ -0,0 +1,127 @@
+from __future__ import annotations
+
+from pathlib import Path
+from typing import TYPE_CHECKING
+from typing import cast
+
+from poetry.core.packages.dependency import Dependency
+from poetry.core.pyproject.toml import PyProjectTOML
+
+from poetry.__version__ import __version__
+from poetry.console.commands.installer_command import InstallerCommand
+from poetry.factory import Factory
+from poetry.packages.project_package import ProjectPackage
+from poetry.poetry import Poetry
+from poetry.utils.env import EnvManager
+from poetry.utils.env import SystemEnv
+from poetry.utils.helpers import directory
+
+
+if TYPE_CHECKING:
+    from poetry.utils.env import Env
+
+
+class SelfCommand(InstallerCommand):
+    ADDITIONAL_PACKAGE_GROUP = "additional"
+
+    @staticmethod
+    def get_default_system_pyproject_file() -> Path:
+        # We separate this out to avoid unwanted side effect during testing while
+        # maintaining dynamic use in help text.
+        #
+        # This is not ideal, but is the simplest solution for now.
+        from poetry.locations import CONFIG_DIR
+
+        return Path(CONFIG_DIR).joinpath("pyproject.toml")
+
+    @property
+    def system_pyproject(self) -> Path:
+        file = self.get_default_system_pyproject_file()
+        file.parent.mkdir(parents=True, exist_ok=True)
+        return file
+
+    def reset_env(self) -> None:
+        self._env = EnvManager.get_system_env(naive=True)
+
+    @property
+    def env(self) -> Env:
+        if self._env is None or not isinstance(self._env, SystemEnv):
+            self.reset_env()
+        return self._env
+
+    @property
+    def default_group(self) -> str:
+        return self.ADDITIONAL_PACKAGE_GROUP
+
+    @property
+    def activated_groups(self) -> set[str]:
+        return {self.default_group}
+
+    def generate_system_pyproject(self) -> None:
+        preserved = {}
+
+        if self.system_pyproject.exists():
+            content = PyProjectTOML(self.system_pyproject).poetry_config
+
+            for key in {"group", "source"}:
+                if key in content:
+                    preserved[key] = content[key]
+
+        package = ProjectPackage(name="poetry-instance", version=__version__)
+        package.add_dependency(Dependency(name="poetry", constraint=f"^{__version__}"))
```

Comment:
I think we shouldn't have a version range for poetry here and pin it instead. We don't know what range - if any - is given by plugin authors and we might end up in situation where a plugin is only available for a small range.

Reply:
I agree. My initial implementation did the same. However, bug fixed in python-poetry/poetry-core#327 causes problems. We can pin it once a new core release with the fix is used.

Reply:
Have you marked this accidentally as resolved? There's still a version range instead of a pinned version. :thinking:

Reply:
Not accidental. Will move to pinned version once a new core version is used in poetry.

Reply:
What's your plan to handle this? Merging now and change it in another PR once a new core release is out? (Hopefully we don't forget it until then :crossed_fingers:)? Or leave this PR open until core is released?

Reply:
I'd rather get the fix done later as the impact of this is not really huge.